src/tactic/declare.lean

```diff
@@ -0,0 +1,194 @@
+/-
+Copyright (c) 2021 Kyle Miller All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import tactic.core
+/-!
+# Canonical declarations
+
+This consists of two user-commands that allow you to define and use
+canonical declarations for certain concepts (for example, all of the
+`variables` to introduce algebraic objects such as a vector space).
+
+It like a parameterized version of `tactic.localized_cmd` (and the code
+is heavily inspired by it).
+
+See the tactic doc entry below for more information.
+-/
+
+open lean lean.parser interactive tactic native
+
+/-- Descriptor for a `def_declare`, constructed when building
+the attribute cache.
+It contains the number of parameters it expects, names for the
+metavariables given in the `def_declare` (for debugging), and
+a function that takes a list of that many parameters and produces
+the code to evaluate. -/
+meta structure canon_decl_def :=
+(nparams : ℕ)
+(default : list string)
+(mk_cmd : list string → tactic string)
+
+/-- The user attribute for canonical declarations. -/
+@[user_attribute]
+meta def canon_decl_attr : user_attribute (rb_lmap name canon_decl_def) unit := {
+  name := "_canon_decl_attr",
+  descr := "(interal) attribute for canonical declarations",
+  parser := failed,
+  cache_cfg := ⟨λ ns, (do dcls ← ns.mmap (λ n, mk_const n >>= eval_expr (name × canon_decl_def)),
+                          return $ rb_lmap.of_list dcls), []⟩ }
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_canon_decls (n : name) (nparams : ℕ) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls.filter (λ decl, decl.nparams = nparams)
+   end
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_all_canon_decls (n : name) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls
+   end
+
+/-- Execute all commands for the given type of canonical declaration. -/
+@[user_command] meta def declare_cmd (_ : parse $ tk "declare") : parser unit :=
+do do_print ← optional (tk "?"),
+   n ← ident,
+   args ← many ident,
+   let args' := args.map name.to_string,
+   decls ← get_canon_decls n args.length,
+   when (decls.length = 0) $
+     fail format!"no canonical declaration for {n} with {args.length} arguments",
+   decls.mmap' $ λ d, do
+     code ← d.mk_cmd args',
+     when do_print.is_some $ trace ("Try this: " ++ code),
+     emit_code_here code -- is it possible to print the code if this fails?
+
+private meta def parse_declare_string (args : list name) (lst : pexpr) :
```

Comment:
I spent a few min thinking if the built in `format!` macro could be used to do this instead, but yeah, it's really tricky. Then I realized it wouldn't be a good idea even then, since we may want to put implicit variables `{...}` in a declare string, which will break the macro.

Reply:
I was originally using `format!` and trying to use `args` as the pattern in a `match` to make the variables available for substitution, but I guess Lean wants to compile the match expression as its own definition, causing errors I didn't want to sort out.

Curly braces are ok in `format!`, but then you have to live with unbalanced curly braces since it'd be `{{x : a}`, and I figured at this point may as well change the template variable symbol.

The reason it's `%(x)` and not `%x`, by the way, is that `ident` consumes spaces, which broke the example of a `def_declare` that includes a `declare`.